File: LimSeoyeon/02_Queue/BOJ_3078_좋은친구.cpp
#include <iostream>
#include <queue>
using namespace std;

int arr[300001];
// 300,000 x 299999 / 2 => 백퍼센트 시간초과.
int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	int N, M;
	cin >> N >> M;
	queue<int> q[21];
	long long res = 0;
	for (int i = 0; i < N; i++) {
		string s;
		cin >> s;
		arr[i] = s.length();
	}
	
	for (int i = 0; i < N; i++) {
		int curQ = arr[i];
		while (!q[curQ].empty()&& i-q[curQ].front()>M) {
			q[curQ].pop();
		}
		res += q[curQ].size();
		q[curQ].push(i);
	}

	cout << res;

}

File: LimSeoyeon/02_Queue/BOJ_3078_좋은친구.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N,M;
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        int [] arr = new int[N];
        Queue<Integer> [] q = new LinkedList[21];
        long res=0;
        for (int i = 0; i < 21; i++) {
            q[i] = new LinkedList<>();
        }
        for(int i=0; i<N; i++){
            String s = br.readLine();
            arr[i] = s.length();
        }

        for(int i=0; i<N; i++){
            int curQ = arr[i];
            while(!q[curQ].isEmpty()&& i-q[curQ].peek()>M){
                q[curQ].poll();
            }
            res+=q[curQ].size();
            q[curQ].add(i);
        }
        System.out.println(res);
    }
}
